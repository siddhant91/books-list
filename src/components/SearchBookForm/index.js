// import clsx from 'clsx';
// import { bool, func, node, oneOf, string } from 'prop-types';
import { useContext } from 'react';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';

// Network
import NetworkUtils from '../../network';

// Components
import TextInput from '../TextInput';

// Contexts
import { AppContext } from '../../contexts/AppContext';
import { BookContext } from '../../contexts/BookContext';

// Styles
import './styles.scss';

function SearchBookForm() {
	const [, setLoaderVisible] = useContext(AppContext);
	const [, setBooksList] = useContext(BookContext);
	const getBooks = async ({ searchText }) => {
		try {
			setLoaderVisible(true);
			const result = await NetworkUtils.makeApiRequest({
				url: '',
				params: {
					q: searchText,
				},
			});

			const { responseData } = result;
			if (responseData) {
				const { items = [] } = responseData;
				if (items.length) {
					setBooksList(items);
				}
			}
		} catch (e) {
			console.log(e);
		} finally {
			setLoaderVisible(false);
		}
	};

	return (
		<div className="bokl-search-form">
			<Formik
				initialValues={{
					searchText: '',
				}}
				validationSchema={Yup.object({
					searchText: Yup.string().trim().required('please input to search'),
				})}
				onSubmit={(values) => {
					getBooks(values);
				}}
			>
				{() => (
					<Form autoComplete="off">
						<TextInput labelKey="Search For Books" name="searchText" placeholder="Search" />
					</Form>
				)}
			</Formik>
		</div>
	);
}

SearchBookForm.propTypes = {};

SearchBookForm.defaultProps = {};

export default SearchBookForm;
