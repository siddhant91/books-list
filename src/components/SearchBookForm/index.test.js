import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';

import SearchBookForm from './index';

afterEach(cleanup);

const renderComponent = (fn) => {
	return <SearchBookForm getBooksList={fn} />;
};
it('Matches snapshot', () => {
	const fn = jest.fn();
	const { asFragment } = render(renderComponent(fn));
	expect(asFragment()).toMatchSnapshot();
});

it('Handles validation of TextInput input element on Submit', async () => {
	const fn = jest.fn();
	const { getByTestId, findByText } = render(renderComponent(fn));
	const button = await findByText('Submit');
	fireEvent.click(button);
	await waitFor(() => {
		expect(fn).toHaveBeenCalledTimes(0);
		const error = getByTestId('TextInput-error');
		expect(error).toBeInTheDocument();
	});
});

it('Handles form Submission', async () => {
	const fn = jest.fn();
	const { getByTestId, findByTestId, findByText } = render(renderComponent(fn));
	const element = getByTestId('TextInput');
	const button = await findByText('Submit');
	fireEvent.change(element, { target: { value: 'testSearch' } });
	const newInput = await findByTestId('TextInput');
	expect(newInput.value).toBe('testSearch');
	fireEvent.click(button);
	await waitFor(() => {});
});
