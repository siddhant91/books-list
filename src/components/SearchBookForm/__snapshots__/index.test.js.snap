// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bokl-search-form"
  >
    <form
      action="#"
      autocomplete="off"
      data-testid="searchBook-form"
    >
      <div
        class="bokl-text-input"
      >
        <label
          class="d-flex bokl-text-input--label"
          for="Search For Books"
        >
          <span>
            <i
              aria-hidden="true"
              class="fa fa-search"
            />
          </span>
          <input
            aria-label="TextInput"
            class=""
            data-testid="TextInput"
            id="Search For Books"
            name="searchText"
            placeholder="Search"
            type="search"
            value=""
          />
        </label>
      </div>
      <button
        class="d-none"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
